# 3 — Model Training (Local GTX 1050 — Optimised)
## FireSpreadNet · Next Day Wildfire Spread

**Hardware target**: NVIDIA GTX 1050 (2–4 GB VRAM)  
**Protocol**: Holdout (train/val/test) — no cross-validation  
**Objective**: Best achievable Dice/IoU on consumer GPU with limited VRAM

### Key optimisations vs. previous FastTrack notebook

| Issue in FastTrack | Fix applied here |
|---|---|
| Only 10 epochs, 5 K samples | Full dataset, 40+ epochs with early stopping |
| 1 tuning trial | 3 structured trials per model |
| PI-CCA skipped entirely | Included with reduced batch & AMP |
| No mixed precision | AMP (torch.amp) for 2× speed + less VRAM |
| No threshold optimisation | Post-training threshold sweep on val set |
| No warmup implemented | Linear warmup for 3 epochs |
| Fixed threshold 0.5 | Optimal threshold per model |
| wind_speed/wind_direction possibly swapped | Verified channel mapping |

### Models compared

| Model | Type | Params | Notes |
|-------|------|--------|-------|
| CA | Physics baseline | 0 | Rothermel rules, no training |
| ConvLSTM | Deep Learning | ~350 K | Lightweight recurrent baseline |
| U-Net + Attention | Deep Learning | ~2.1 M | Dense segmentation |
| PI-CCA | Hybrid (ours) | ~1.5 M | Physics-informed, main contribution |

In [1]:
import gc, json, os, sys, time, warnings
from copy import deepcopy
from pathlib import Path

import numpy as np
import matplotlib.pyplot as plt
import matplotlib as mpl
import seaborn as sns
import pandas as pd                        # ★ FIX: was missing

import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import DataLoader, Subset
from torch.amp import autocast
from torch.cuda.amp import GradScaler

warnings.filterwarnings('ignore', category=FutureWarning)
sns.set_theme(style='whitegrid', font_scale=1.1)
mpl.rcParams.update({'savefig.dpi': 200, 'figure.dpi': 100})
%matplotlib inline

sys.path.insert(0, os.path.abspath('..'))

from config import MODEL_CONFIG, TRAIN_CONFIG, SEED, MODELS_DIR, RESULTS_DIR, FIGURES_DIR
from src.data.dataset import get_dataloaders, FireSpreadDataset
from src.models.cellular_automata import CellularAutomataModel
from src.models.convlstm import ConvLSTMModel
from src.models.unet import UNetFire
from src.models.pi_cca import PIConvCellularAutomaton
from src.training.trainer import FocalLoss, DiceLoss, CombinedLoss, compute_metrics

_cfg_path = Path().resolve() / 'setup_config.json'
if not _cfg_path.exists():
    raise FileNotFoundError('setup_config.json not found --- run 00_Setup.ipynb first')
cfg = json.load(open(_cfg_path))

PROCESSED_DIR    = Path(cfg['PROCESSED_DIR'])
FEATURE_CHANNELS = cfg['FEATURE_CHANNELS']
N_INPUT_CHANNELS = cfg['N_INPUT_CHANNELS']
CH               = cfg['CH']
norm_stats       = cfg['norm_stats']

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
if torch.cuda.is_available():
    gpu_name = torch.cuda.get_device_name(0)
    vram_gb  = torch.cuda.get_device_properties(0).total_memory / (1024**3)
    print(f'GPU: {gpu_name} ({vram_gb:.1f} GB VRAM)')
else:
    vram_gb = 0
    print('CPU mode (no GPU detected)')

torch.manual_seed(SEED)
np.random.seed(SEED)
if torch.cuda.is_available():
    torch.cuda.manual_seed_all(SEED)
    torch.backends.cudnn.deterministic = False
    torch.backends.cudnn.benchmark = True

print(f'Device: {device} | Seed: {SEED}')


GPU: NVIDIA GeForce GTX 1050 (3.0 GB VRAM)
Device: cuda | Seed: 42


## 3.1 Data Loading & Verification

In [2]:
# ==========================================
# Cell 2 - Data Loading (full dataset, GTX 1050 batch sizes)
# ==========================================

BATCH_SIZES = {'convlstm': 8, 'unet': 8, 'pi_cca': 4, 'ca': 16}

# num_workers=0 for Windows Jupyter compatibility
loaders = get_dataloaders(
    PROCESSED_DIR, batch_size=8, num_workers=0,
    augment_train=True, seed=SEED, stats=norm_stats,
)

for name, loader in loaders.items():
    print(f'{name}: {len(loader.dataset)} samples, {len(loader)} batches')

x_batch, y_batch = next(iter(loaders['train']))
print(f'\nBatch: X={x_batch.shape}, Y={y_batch.shape}')
print(f'Y unique values: {torch.unique(y_batch).tolist()}')

fire_pct = y_batch.float().mean().item() * 100
print(f'Fire pixel ratio: {fire_pct:.2f}%')
if fire_pct < 2.0:
    print('  WARNING: Severe imbalance (<2%) - focal_alpha=0.90 applied')

# IMPORTANT: after preprocessing fix, all channels should be in [-5, 5]
print('\n-- Post-normalisation range check (should be in [-5, 5]) --')
all_ok = True
for ch_idx, ch_name in enumerate(FEATURE_CHANNELS):
    ch = x_batch[:, ch_idx]
    mn, mx = ch.min().item(), ch.max().item()
    flag = '[OUTLIER!]' if abs(mn) > 5.5 or abs(mx) > 5.5 else '[OK]'
    if abs(mn) > 5.5 or abs(mx) > 5.5:
        all_ok = False
    print(f'  {ch_name:>18s}: [{mn:+.2f}, {mx:+.2f}]  {flag}')

if all_ok:
    print('\nAll channels normalised correctly - training should be stable.')
else:
    print('\nERROR: Extreme values remain - check src/data/preprocessing.py')


train: 14979 samples, 1872 batches
val: 1877 samples, 235 batches
test: 1689 samples, 212 batches

Batch: X=torch.Size([8, 12, 64, 64]), Y=torch.Size([8, 1, 64, 64])
Y unique values: [0.0, 1.0]
Fire pixel ratio: 1.01%

-- Post-normalisation range check (should be in [-5, 5]) --
           elevation: [-1.05, +3.02]  [OK]
          wind_speed: [-0.02, +0.05]  [OK]
      wind_direction: [-1.23, +1.25]  [OK]
            min_temp: [-0.94, +0.70]  [OK]
            max_temp: [-0.71, +0.78]  [OK]
            humidity: [-1.09, +0.86]  [OK]
       precipitation: [-0.23, +0.05]  [OK]
       drought_index: [-2.31, +2.83]  [OK]
                ndvi: [-4.09, +1.65]  [OK]
                 erc: [-1.26, +1.66]  [OK]
          population: [-0.65, +4.80]  [OK]
      prev_fire_mask: [+0.00, +1.00]  [OK]

All channels normalised correctly - training should be stable.


## 3.2 Enhanced Training Infrastructure

Key improvements over the base `Trainer`:
- **Mixed precision** (AMP) for ~2× speedup and ~50% VRAM saving
- **Linear warmup** followed by cosine annealing
- **Gradient accumulation** for effective larger batch sizes
- **Optimal threshold search** on validation set
- **Comprehensive logging** with per-epoch timing

In [3]:

# ══════════════════════════════════════════════════════════════
# Cell 3 — Enhanced Trainer with AMP + Warmup + Threshold Opt
#          ★ FIXED: running-sum metrics to prevent GPU OOM ★
#          ★ FIXED: focal_alpha=0.99 (was 0.75 → caused collapse) ★
# ══════════════════════════════════════════════════════════════
import gc
from tqdm.auto import tqdm

def compute_metrics_at_threshold(pred, target, threshold=0.5):
    with torch.no_grad():
        p_bin = (pred > threshold).float()
        t_bin = target.float()
        tp = (p_bin * t_bin).sum().item()
        fp = (p_bin * (1 - t_bin)).sum().item()
        fn = ((1 - p_bin) * t_bin).sum().item()
        tn = ((1 - p_bin) * (1 - t_bin)).sum().item()
        eps = 1e-8
        precision   = tp / (tp + fp + eps)
        recall      = tp / (tp + fn + eps)
        f1          = 2 * precision * recall / (precision + recall + eps)
        iou         = tp / (tp + fp + fn + eps)
        dice        = 2 * tp / (2 * tp + fp + fn + eps)
        accuracy    = (tp + tn) / (tp + fp + fn + tn + eps)
        specificity = tn / (tn + fp + eps)
    return {'iou': iou, 'dice': dice, 'f1': f1, 'precision': precision,
            'recall': recall, 'accuracy': accuracy, 'specificity': specificity,
            'threshold': threshold}

def find_optimal_threshold(preds, targets, thresholds=None):
    if thresholds is None:
        thresholds = np.arange(0.05, 0.90, 0.05)
    best_dice, best_t = 0, 0.5
    for t in thresholds:
        m = compute_metrics_at_threshold(preds, targets, float(t))
        if m['dice'] > best_dice:
            best_dice, best_t = m['dice'], float(t)
    return best_t, best_dice


def _compute_running_metrics(tp, fp, fn, tn):
    """Compute metrics from running confusion matrix sums."""
    eps = 1e-8
    precision   = tp / (tp + fp + eps)
    recall      = tp / (tp + fn + eps)
    f1          = 2 * precision * recall / (precision + recall + eps)
    iou         = tp / (tp + fp + fn + eps)
    dice        = 2 * tp / (2 * tp + fp + fn + eps)
    accuracy    = (tp + tn) / (tp + fp + fn + tn + eps)
    specificity = tn / (tn + fp + eps)
    return {'iou': iou, 'dice': dice, 'f1': f1, 'precision': precision,
            'recall': recall, 'accuracy': accuracy, 'specificity': specificity,
            'auc_pr': 0.0}


class EnhancedTrainer:
    """Training loop optimised for GTX 1050: AMP, warmup, threshold opt.

    ★ FIX: uses running-sum confusion matrix instead of accumulating
      all predictions — saves ~500 MB of GPU memory per epoch.
    ★ FIX: default focal_alpha=0.99 (= 1 - fire_freq for ~1% fire rate).
      alpha_fire × N_fire = alpha_no_fire × N_no_fire  ↔  alpha ≈ 0.989.
    """

    def __init__(self, model, model_name, device, config, use_amp=True, accum_steps=1):
        self.model       = model.to(device)
        self.model_name  = model_name
        self.device      = device
        self.cfg         = config
        self.use_amp     = use_amp and torch.cuda.is_available()
        self.accum_steps = accum_steps
        # ★ FIX: default alpha raised from 0.75 → 0.99
        self.criterion   = CombinedLoss(
            focal_w=config.get('focal_weight', 0.5),
            dice_w=config.get('dice_weight', 0.5),
            alpha=config.get('focal_alpha', 0.99),
            gamma=config.get('focal_gamma', 2.0),
        )
        params = [p for p in model.parameters() if p.requires_grad]
        self.has_params = len(params) > 0
        if self.has_params:
            self.optimizer = torch.optim.AdamW(
                params,
                lr=config.get('learning_rate', 3e-4),
                weight_decay=config.get('weight_decay', 1e-4),
            )
        else:
            self.optimizer = None
        self.scaler  = GradScaler() if self.use_amp else None
        self.history = {'train_loss': [], 'val_loss': [],
                        'train_metrics': [], 'val_metrics': [],
                        'lr': [], 'epoch_time': []}

    def _get_lr(self, epoch, total_epochs, warmup_epochs=3):
        base_lr = self.cfg.get('learning_rate', 3e-4)
        min_lr  = 1e-6
        if epoch < warmup_epochs:
            return base_lr * (epoch + 1) / warmup_epochs
        progress = (epoch - warmup_epochs) / max(1, total_epochs - warmup_epochs)
        return min_lr + 0.5 * (base_lr - min_lr) * (1 + np.cos(np.pi * progress))

    def train(self, train_loader, val_loader):
        epochs    = self.cfg.get('epochs', 40)
        patience  = self.cfg.get('early_stopping_patience', 10)
        warmup    = self.cfg.get('warmup_epochs', 3)
        grad_clip = self.cfg.get('gradient_clip', 1.0)
        best_val_dice, no_improve = 0.0, 0
        save_dir = MODELS_DIR / self.model_name
        save_dir.mkdir(parents=True, exist_ok=True)
        alpha_used = self.cfg.get('focal_alpha', 0.99)
        print(f'\n{"="*65}')
        print(f'Training {self.model_name} | {epochs} epochs | device={self.device}')
        print(f'AMP={self.use_amp} | accum_steps={self.accum_steps} | patience={patience}')
        print(f'focal_alpha={alpha_used:.3f} | focal_gamma={self.cfg.get("focal_gamma",2.0)}')
        print(f'Expected gradient balance: fire={alpha_used:.3f} × N_fire ≈ no-fire={(1-alpha_used):.3f} × N_no-fire')
        print(f'{"="*65}')
        for epoch in range(epochs):
            t0 = time.time()
            lr = self._get_lr(epoch, epochs, warmup)
            if self.optimizer:
                for pg in self.optimizer.param_groups:
                    pg['lr'] = lr
            if self.has_params:
                train_loss, train_metrics = self._train_epoch(train_loader, grad_clip)
            else:
                train_loss, train_metrics = self._eval_epoch(train_loader)
            val_loss, val_metrics = self._eval_epoch(val_loader)
            dt = time.time() - t0
            self.history['train_loss'].append(float(train_loss))
            self.history['val_loss'].append(float(val_loss))
            self.history['train_metrics'].append(train_metrics)
            self.history['val_metrics'].append(val_metrics)
            self.history['lr'].append(float(lr))
            self.history['epoch_time'].append(float(dt))
            # ★ Early collapse detection: recall should never drop to 0
            recall_flag = ' ⚠ RECALL=0 CHECK ALPHA!' if train_metrics['recall'] < 0.01 and epoch >= 2 else ''
            print(f'Epoch {epoch+1:3d}/{epochs} | '
                  f'Loss: {train_loss:.4f}/{val_loss:.4f} | '
                  f'Dice: {train_metrics["dice"]:.4f}/{val_metrics["dice"]:.4f} | '
                  f'Recall: {val_metrics["recall"]:.3f} | '
                  f'LR: {lr:.2e} | {dt:.1f}s{recall_flag}')
            if val_metrics['dice'] > best_val_dice:
                best_val_dice = val_metrics['dice']
                no_improve    = 0
                torch.save(self.model.state_dict(), save_dir / 'best_model.pt')
            else:
                no_improve += 1
            if no_improve >= patience:
                print(f'Early stopping at epoch {epoch+1} (best Dice={best_val_dice:.4f})')
                break
            if torch.cuda.is_available():
                torch.cuda.empty_cache()
            gc.collect()
        with open(save_dir / 'training_history.json', 'w') as f:
            json.dump(self.history, f, indent=2, default=str)
        print(f'Best val Dice: {best_val_dice:.4f}')
        return self.history

    def _train_epoch(self, loader, grad_clip=1.0):
        """★ FIX: running-sum metrics — no pred accumulation."""
        self.model.train()
        total_loss = 0
        tp_sum = fp_sum = fn_sum = tn_sum = 0
        self.optimizer.zero_grad()
        for step, (x, y) in enumerate(tqdm(loader, desc='Train', leave=False)):
            x, y = x.to(self.device, non_blocking=True), y.to(self.device, non_blocking=True)
            if self.use_amp:
                with autocast('cuda'):
                    pred = self.model(x)
                # ★ Loss computed in FP32 (outside autocast) for numerical stability
                loss = self.criterion(pred.float(), y.float()) / self.accum_steps
                self.scaler.scale(loss).backward()
                if (step + 1) % self.accum_steps == 0:
                    self.scaler.unscale_(self.optimizer)
                    nn.utils.clip_grad_norm_(self.model.parameters(), grad_clip)
                    self.scaler.step(self.optimizer)
                    self.scaler.update()
                    self.optimizer.zero_grad()
            else:
                pred = self.model(x)
                loss = self.criterion(pred, y) / self.accum_steps
                loss.backward()
                if (step + 1) % self.accum_steps == 0:
                    nn.utils.clip_grad_norm_(self.model.parameters(), grad_clip)
                    self.optimizer.step()
                    self.optimizer.zero_grad()
            total_loss += loss.item() * self.accum_steps * x.size(0)
            with torch.no_grad():
                p_bin = (pred.detach() > 0.5).float()
                t_bin = y.float()
                tp_sum += (p_bin * t_bin).sum().item()
                fp_sum += (p_bin * (1 - t_bin)).sum().item()
                fn_sum += ((1 - p_bin) * t_bin).sum().item()
                tn_sum += ((1 - p_bin) * (1 - t_bin)).sum().item()
        avg_loss = total_loss / len(loader.dataset)
        metrics  = _compute_running_metrics(tp_sum, fp_sum, fn_sum, tn_sum)
        return avg_loss, metrics

    @torch.no_grad()
    def _eval_epoch(self, loader):
        """★ FIX: running-sum metrics — no pred accumulation."""
        self.model.eval()
        total_loss = 0
        tp_sum = fp_sum = fn_sum = tn_sum = 0
        for x, y in tqdm(loader, desc='Eval', leave=False):
            x, y = x.to(self.device, non_blocking=True), y.to(self.device, non_blocking=True)
            if self.use_amp:
                with autocast('cuda'):
                    pred = self.model(x)
                loss = self.criterion(pred.float(), y.float())
            else:
                pred = self.model(x)
                loss = self.criterion(pred, y)
            total_loss += loss.item() * x.size(0)
            p_bin = (pred > 0.5).float()
            t_bin = y.float()
            tp_sum += (p_bin * t_bin).sum().item()
            fp_sum += (p_bin * (1 - t_bin)).sum().item()
            fn_sum += ((1 - p_bin) * t_bin).sum().item()
            tn_sum += ((1 - p_bin) * (1 - t_bin)).sum().item()
        avg_loss = total_loss / len(loader.dataset)
        metrics  = _compute_running_metrics(tp_sum, fp_sum, fn_sum, tn_sum)
        return avg_loss, metrics

    @torch.no_grad()
    def evaluate_with_threshold_opt(self, val_loader, test_loader=None):
        """Threshold optimisation — uses CPU for pred storage."""
        self.model.eval()
        val_preds, val_targets = [], []
        for x, y in val_loader:
            x = x.to(self.device, non_blocking=True)
            if self.use_amp:
                with autocast('cuda'):
                    pred = self.model(x)
            else:
                pred = self.model(x)
            val_preds.append(pred.float().cpu())
            val_targets.append(y.float())
        val_preds   = torch.cat(val_preds)
        val_targets = torch.cat(val_targets)
        best_t, best_dice = find_optimal_threshold(val_preds, val_targets)
        val_metrics = compute_metrics_at_threshold(val_preds, val_targets, best_t)
        print(f'  Optimal threshold: {best_t:.2f} (val Dice={best_dice:.4f})')
        results = {'val': val_metrics, 'optimal_threshold': best_t}
        if test_loader is not None:
            test_preds, test_targets = [], []
            for x, y in test_loader:
                x = x.to(self.device, non_blocking=True)
                if self.use_amp:
                    with autocast('cuda'):
                        pred = self.model(x)
                else:
                    pred = self.model(x)
                test_preds.append(pred.float().cpu())
                test_targets.append(y.float())
            results['test'] = compute_metrics_at_threshold(
                torch.cat(test_preds), torch.cat(test_targets), best_t)
        return results

print('EnhancedTrainer ready (memory-optimised, alpha=0.99 corrected)')


EnhancedTrainer ready (memory-optimised, alpha=0.99 corrected)


## 3.3 Model Definitions & Hyperparameter Search Space

In [ ]:
# ══════════════════════════════════════════════════════════════
# Cell 4 — Model registry & Hyperparameter grid (GTX 1050)
# ══════════════════════════════════════════════════════════════
from torch.utils.data import WeightedRandomSampler

MODEL_CLASSES = {
    'ca': CellularAutomataModel,
    'convlstm': ConvLSTMModel,
    'unet': UNetFire,
    'pi_cca': PIConvCellularAutomaton,
}

# FIX: reduce U-Net from 7.8M to ~2M params to fit in 3GB VRAM
MODEL_CONFIG['unet']['base_filters'] = 16

print('Model parameter counts:')
for name, cls in MODEL_CLASSES.items():
    m = cls(config=MODEL_CONFIG[name])
    n_p = sum(p.numel() for p in m.parameters() if p.requires_grad)
    print(f'  {name:>10s}: {n_p:>10,} params')
    del m

# ══════════════════════════════════════════════════════════════
# ★★ CRITICAL FIX: focal_alpha calibrated for 1.07% fire rate ★★
# ══════════════════════════════════════════════════════════════
# Math: for equal pixel-level gradient contribution from fire vs no-fire:
#   alpha_fire × N_fire  =  alpha_no_fire × N_no_fire
#   alpha_fire × 0.0107  =  (1 - alpha_fire) × 0.9893
#   → alpha_fire = 0.9893 ≈ 0.99
#
# The OLD values (0.85–0.95) caused the gradient to be dominated by
# no-fire pixels (ratio ≈ 30:1), causing trivial collapse to all-no-fire.
HP_GRID = {
    'convlstm': [
        {'learning_rate': 3e-4, 'weight_decay': 1e-4, 'focal_alpha': 0.99,  'focal_gamma': 2.0, 'dropout': 0.15},
        {'learning_rate': 1e-4, 'weight_decay': 5e-5, 'focal_alpha': 0.98,  'focal_gamma': 2.5, 'dropout': 0.10},
        {'learning_rate': 5e-4, 'weight_decay': 1e-4, 'focal_alpha': 0.995, 'focal_gamma': 3.0, 'dropout': 0.20},
    ],
    'unet': [
        {'learning_rate': 3e-4, 'weight_decay': 1e-4, 'focal_alpha': 0.99,  'focal_gamma': 2.0, 'dropout': 0.15},
        {'learning_rate': 1e-4, 'weight_decay': 5e-5, 'focal_alpha': 0.98,  'focal_gamma': 2.5, 'dropout': 0.10},
        {'learning_rate': 5e-4, 'weight_decay': 5e-5, 'focal_alpha': 0.995, 'focal_gamma': 3.0, 'dropout': 0.20},
    ],
    'pi_cca': [
        {'learning_rate': 3e-4, 'weight_decay': 1e-4, 'focal_alpha': 0.99,  'focal_gamma': 2.0, 'dropout': 0.15},
        {'learning_rate': 1e-4, 'weight_decay': 5e-5, 'focal_alpha': 0.98,  'focal_gamma': 2.5, 'dropout': 0.10},
        {'learning_rate': 2e-4, 'weight_decay': 1e-4, 'focal_alpha': 0.995, 'focal_gamma': 3.0, 'dropout': 0.15},
    ],
}

TUNING_EPOCHS = 6
FINAL_EPOCHS  = 30
PATIENCE      = 8

candidate_models = ['convlstm', 'unet', 'pi_cca']

print(f'Training plan: {candidate_models}')
print(f'Tuning: {TUNING_EPOCHS} epochs × {len(HP_GRID["unet"])} trials per model')
print(f'Final:  {FINAL_EPOCHS} epochs with early stopping (patience={PATIENCE})')

Model parameter counts:
          ca:          0 params
    convlstm:    418,913 params
        unet:  1,966,485 params
      pi_cca:    269,431 params
Training plan: ['convlstm', 'unet', 'pi_cca']
Tuning: 6 epochs × 3 trials per model
Final:  30 epochs with early stopping (patience=8)



# ══════════════════════════════════════════════════════════════
# Cell 5 — Main training loop (GPU-optimised, imbalance-corrected)
# ══════════════════════════════════════════════════════════════

all_results = {}
tuning_summary = {}

# ★ FIX: Pre-compute fire weights ONCE for all models (avoid 15K dataset[i] calls per model)
print('Pre-computing fire weights for sampler (one-time cost)...')
def _get_fire_weights_cached(dataset):
    """Compute fire weights efficiently. 
    
    Cache result to avoid recomputing for every model (was causing U-Net hang).
    """
    cache_key = id(dataset)
    if not hasattr(_get_fire_weights_cached, '_cache'):
        _get_fire_weights_cached._cache = {}
    
    if cache_key in _get_fire_weights_cached._cache:
        return _get_fire_weights_cached._cache[cache_key]
    
    # Try fast path first
    if hasattr(dataset, 'targets') and hasattr(dataset, 'indices'):
        raw_targets = dataset.targets[dataset.indices]        # (N, 1, H, W)
        fire_counts = raw_targets.mean(axis=(1, 2, 3))        # (N,) fire fraction
    else:
        # Slow path: only compute once and cache
        print(f'  Loading {len(dataset)} samples to compute fire statistics...')
        fire_counts = np.array([float(dataset[i][1].mean()) for i in tqdm(range(len(dataset)), desc='Fire count', leave=False)])
    
    weights = np.maximum(fire_counts, 1e-3).astype(np.float64)
    weights = weights / weights.sum()
    
    result = (weights, fire_counts)
    _get_fire_weights_cached._cache[cache_key] = result
    return result

weights_cached, fire_counts = _get_fire_weights_cached(loaders['train'].dataset)
n_fire_samples = (fire_counts > 0).sum()
print(f'✓ Fire weights computed: {n_fire_samples}/{len(loaders["train"].dataset)} samples with fire | '
      f'mean fire density {fire_counts.mean()*100:.3f}%')

for model_name in candidate_models:
    print(f'\n{"═"*70}')
    print(f'  MODEL: {MODEL_CONFIG[model_name]["name"]}')
    print(f'{"═"*70}')

    # Model-specific batch size & accumulation
    bs = BATCH_SIZES.get(model_name, 8)
    accum = 2 if model_name == 'pi_cca' else 1   # effective bs=8 for PI-CCA

    # ★ Use fire-weighted sampler for training (now cached — instant for U-Net+)
    generator = torch.Generator().manual_seed(SEED)
    sampler = WeightedRandomSampler(
        weights=torch.from_numpy(weights_cached),
        num_samples=len(loaders['train'].dataset),
        replacement=True,
        generator=generator,
    )
    train_loader = DataLoader(
        loaders['train'].dataset, batch_size=bs, sampler=sampler,
        num_workers=0, pin_memory=True, drop_last=True,
    )
    val_loader = DataLoader(
        loaders['val'].dataset, batch_size=bs, shuffle=False,
        num_workers=0, pin_memory=True,
    )

    # ── Phase 1: Hyperparameter tuning ──
    trial_results = []
    for trial_idx, hp in enumerate(HP_GRID[model_name]):
        print(f'\n  Trial {trial_idx+1}/{len(HP_GRID[model_name])}: {hp}')

        mcfg = deepcopy(MODEL_CONFIG[model_name])
        mcfg['dropout'] = hp.get('dropout', mcfg.get('dropout', 0.2))
        model = MODEL_CLASSES[model_name](config=mcfg).to(device)

        tcfg = {
            'epochs': TUNING_EPOCHS,
            'learning_rate': hp['learning_rate'],
            'weight_decay': hp['weight_decay'],
            'focal_alpha': hp['focal_alpha'],
            'focal_gamma': hp['focal_gamma'],
            'focal_weight': 0.5,
            'dice_weight': 0.5,
            'early_stopping_patience': TUNING_EPOCHS,  # no early stop during tuning
            'gradient_clip': 1.0,
            'warmup_epochs': 2,
        }

        trainer = EnhancedTrainer(
            model, model_name, device, tcfg,
            use_amp=True, accum_steps=accum,
        )
        hist = trainer.train(train_loader, val_loader)

        val_dices = [m['dice'] for m in hist['val_metrics']]
        best_dice = max(val_dices)
        trial_results.append({
            'hp': hp,
            'best_dice': float(best_dice),
            'last_dice': float(val_dices[-1]),
            'best_epoch': int(np.argmax(val_dices)) + 1,
        })
        print(f'    Best val Dice: {best_dice:.4f} (epoch {np.argmax(val_dices)+1})')

        del model, trainer
        if torch.cuda.is_available():
            torch.cuda.empty_cache()
        gc.collect()

    # Select best trial
    best_trial = max(trial_results, key=lambda x: x['best_dice'])
    best_hp = best_trial['hp']
    tuning_summary[model_name] = {
        'selection_metric': 'dice',
        'best_trial': best_trial,
        'all_trials': trial_results,
    }
    print(f'\n  ★ Best HP: {best_hp} (Dice={best_trial["best_dice"]:.4f})')

    # ── Phase 2: Final training with best config ──
    print(f'\n  Final training: {FINAL_EPOCHS} epochs...')
    mcfg = deepcopy(MODEL_CONFIG[model_name])
    mcfg['dropout'] = best_hp.get('dropout', mcfg.get('dropout', 0.2))
    model = MODEL_CLASSES[model_name](config=mcfg).to(device)

    tcfg = {
        'epochs': FINAL_EPOCHS,
        'learning_rate': best_hp['learning_rate'],
        'weight_decay': best_hp['weight_decay'],
        'focal_alpha': best_hp['focal_alpha'],
        'focal_gamma': best_hp['focal_gamma'],
        'focal_weight': 0.5,
        'dice_weight': 0.5,
        'early_stopping_patience': PATIENCE,
        'gradient_clip': 1.0,
        'warmup_epochs': 3,
    }

    trainer = EnhancedTrainer(
        model, model_name, device, tcfg,
        use_amp=True, accum_steps=accum,
    )
    history = trainer.train(train_loader, val_loader)

    # ── Phase 3: Threshold optimisation + test evaluation ──
    ckpt_path = MODELS_DIR / model_name / 'best_model.pt'
    model.load_state_dict(torch.load(ckpt_path, map_location=device, weights_only=True))

    test_loader = DataLoader(
        loaders['test'].dataset, batch_size=bs, shuffle=False,
        num_workers=0, pin_memory=True,
    )
    eval_results = trainer.evaluate_with_threshold_opt(val_loader, test_loader)

    all_results[model_name] = {
        'history': history,
        'tuning': tuning_summary[model_name],
        'eval': eval_results,
    }

    # Save
    save_dir = MODELS_DIR / model_name
    with open(save_dir / 'hyperparameter_tuning.json', 'w') as f:
        json.dump(tuning_summary[model_name], f, indent=2, default=str)
    with open(save_dir / 'eval_results.json', 'w') as f:
        json.dump(eval_results, f, indent=2, default=str)

    del model, trainer
    if torch.cuda.is_available():
        torch.cuda.empty_cache()
    gc.collect()

# ── Physics baseline (CA) ──
print(f'\n{"═"*70}')
print('  MODEL: Cellular Automata (physics baseline — no training)')
print(f'{"═"*70}')

ca_loader_val  = DataLoader(loaders['val'].dataset,  batch_size=16, shuffle=False, num_workers=0, pin_memory=True)
ca_loader_test = DataLoader(loaders['test'].dataset, batch_size=16, shuffle=False, num_workers=0, pin_memory=True)
ca_model = CellularAutomataModel(config=MODEL_CONFIG['ca']).to(device)
ca_trainer = EnhancedTrainer(ca_model, 'ca', device, TRAIN_CONFIG, use_amp=False)
ca_eval = ca_trainer.evaluate_with_threshold_opt(ca_loader_val, ca_loader_test)
all_results['ca'] = {'eval': ca_eval}

print('\n✅ All models trained successfully!')


In [ ]:

# ══════════════════════════════════════════════════════════════
# Cell 5 — Main training loop (GPU-optimised, imbalance-corrected)
# ══════════════════════════════════════════════════════════════

all_results = {}
tuning_summary = {}

for model_name in candidate_models:
    print(f'\n{"═"*70}')
    print(f'  MODEL: {MODEL_CONFIG[model_name]["name"]}')
    print(f'{"═"*70}')

    # Model-specific batch size & accumulation
    bs = BATCH_SIZES.get(model_name, 8)
    accum = 2 if model_name == 'pi_cca' else 1   # effective bs=8 for PI-CCA

    # ★ Use fire-weighted sampler for training to enrich fire-pixel batches
    train_loader = make_fire_weighted_loader(
        loaders['train'].dataset, batch_size=bs, num_workers=0, drop_last=True, seed=SEED
    )
    val_loader = DataLoader(
        loaders['val'].dataset, batch_size=bs, shuffle=False,
        num_workers=0, pin_memory=True,
    )

    # ── Phase 1: Hyperparameter tuning ──
    trial_results = []
    for trial_idx, hp in enumerate(HP_GRID[model_name]):
        print(f'\n  Trial {trial_idx+1}/{len(HP_GRID[model_name])}: {hp}')

        mcfg = deepcopy(MODEL_CONFIG[model_name])
        mcfg['dropout'] = hp.get('dropout', mcfg.get('dropout', 0.2))
        model = MODEL_CLASSES[model_name](config=mcfg).to(device)

        tcfg = {
            'epochs': TUNING_EPOCHS,
            'learning_rate': hp['learning_rate'],
            'weight_decay': hp['weight_decay'],
            'focal_alpha': hp['focal_alpha'],
            'focal_gamma': hp['focal_gamma'],
            'focal_weight': 0.5,
            'dice_weight': 0.5,
            'early_stopping_patience': TUNING_EPOCHS,  # no early stop during tuning
            'gradient_clip': 1.0,
            'warmup_epochs': 2,
        }

        trainer = EnhancedTrainer(
            model, model_name, device, tcfg,
            use_amp=True, accum_steps=accum,
        )
        hist = trainer.train(train_loader, val_loader)

        val_dices = [m['dice'] for m in hist['val_metrics']]
        best_dice = max(val_dices)
        trial_results.append({
            'hp': hp,
            'best_dice': float(best_dice),
            'last_dice': float(val_dices[-1]),
            'best_epoch': int(np.argmax(val_dices)) + 1,
        })
        print(f'    Best val Dice: {best_dice:.4f} (epoch {np.argmax(val_dices)+1})')

        del model, trainer
        if torch.cuda.is_available():
            torch.cuda.empty_cache()
        gc.collect()

    # Select best trial
    best_trial = max(trial_results, key=lambda x: x['best_dice'])
    best_hp = best_trial['hp']
    tuning_summary[model_name] = {
        'selection_metric': 'dice',
        'best_trial': best_trial,
        'all_trials': trial_results,
    }
    print(f'\n  ★ Best HP: {best_hp} (Dice={best_trial["best_dice"]:.4f})')

    # ── Phase 2: Final training with best config ──
    print(f'\n  Final training: {FINAL_EPOCHS} epochs...')
    mcfg = deepcopy(MODEL_CONFIG[model_name])
    mcfg['dropout'] = best_hp.get('dropout', mcfg.get('dropout', 0.2))
    model = MODEL_CLASSES[model_name](config=mcfg).to(device)

    tcfg = {
        'epochs': FINAL_EPOCHS,
        'learning_rate': best_hp['learning_rate'],
        'weight_decay': best_hp['weight_decay'],
        'focal_alpha': best_hp['focal_alpha'],
        'focal_gamma': best_hp['focal_gamma'],
        'focal_weight': 0.5,
        'dice_weight': 0.5,
        'early_stopping_patience': PATIENCE,
        'gradient_clip': 1.0,
        'warmup_epochs': 3,
    }

    trainer = EnhancedTrainer(
        model, model_name, device, tcfg,
        use_amp=True, accum_steps=accum,
    )
    history = trainer.train(train_loader, val_loader)

    # ── Phase 3: Threshold optimisation + test evaluation ──
    ckpt_path = MODELS_DIR / model_name / 'best_model.pt'
    model.load_state_dict(torch.load(ckpt_path, map_location=device, weights_only=True))

    test_loader = DataLoader(
        loaders['test'].dataset, batch_size=bs, shuffle=False,
        num_workers=0, pin_memory=True,
    )
    eval_results = trainer.evaluate_with_threshold_opt(val_loader, test_loader)

    all_results[model_name] = {
        'history': history,
        'tuning': tuning_summary[model_name],
        'eval': eval_results,
    }

    # Save
    save_dir = MODELS_DIR / model_name
    with open(save_dir / 'hyperparameter_tuning.json', 'w') as f:
        json.dump(tuning_summary[model_name], f, indent=2, default=str)
    with open(save_dir / 'eval_results.json', 'w') as f:
        json.dump(eval_results, f, indent=2, default=str)

    del model, trainer
    if torch.cuda.is_available():
        torch.cuda.empty_cache()
    gc.collect()

# ── Physics baseline (CA) ──
print(f'\n{"═"*70}')
print('  MODEL: Cellular Automata (physics baseline — no training)')
print(f'{"═"*70}')

ca_loader_val  = DataLoader(loaders['val'].dataset,  batch_size=16, shuffle=False, num_workers=0, pin_memory=True)
ca_loader_test = DataLoader(loaders['test'].dataset, batch_size=16, shuffle=False, num_workers=0, pin_memory=True)
ca_model = CellularAutomataModel(config=MODEL_CONFIG['ca']).to(device)
ca_trainer = EnhancedTrainer(ca_model, 'ca', device, TRAIN_CONFIG, use_amp=False)
ca_eval = ca_trainer.evaluate_with_threshold_opt(ca_loader_val, ca_loader_test)
all_results['ca'] = {'eval': ca_eval}

print('\n✅ All models trained successfully!')



══════════════════════════════════════════════════════════════════════
  MODEL: ConvLSTM
══════════════════════════════════════════════════════════════════════


  WeightedSampler: 13273/14979 samples with fire | mean fire density 1.073%

  Trial 1/3: {'learning_rate': 0.0003, 'weight_decay': 0.0001, 'focal_alpha': 0.99, 'focal_gamma': 2.0, 'dropout': 0.15}

Training convlstm | 6 epochs | device=cuda
AMP=True | accum_steps=1 | patience=6
focal_alpha=0.990 | focal_gamma=2.0
Expected gradient balance: fire=0.990 × N_fire ≈ no-fire=0.010 × N_no-fire


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   1/6 | Loss: 0.4058/0.4545 | Dice: 0.2236/0.1180 | Recall: 0.467 | LR: 1.50e-04 | 152.7s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   2/6 | Loss: 0.3688/0.4270 | Dice: 0.3529/0.1506 | Recall: 0.448 | LR: 3.00e-04 | 149.6s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   3/6 | Loss: 0.3594/0.4144 | Dice: 0.3793/0.2665 | Recall: 0.421 | LR: 3.00e-04 | 150.0s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   4/6 | Loss: 0.3572/0.4206 | Dice: 0.3848/0.2402 | Recall: 0.424 | LR: 2.56e-04 | 150.8s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   5/6 | Loss: 0.3533/0.4165 | Dice: 0.3932/0.2480 | Recall: 0.439 | LR: 1.50e-04 | 150.4s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   6/6 | Loss: 0.3516/0.4122 | Dice: 0.3964/0.2460 | Recall: 0.431 | LR: 4.48e-05 | 150.6s
Best val Dice: 0.2665
    Best val Dice: 0.2665 (epoch 3)

  Trial 2/3: {'learning_rate': 0.0001, 'weight_decay': 5e-05, 'focal_alpha': 0.98, 'focal_gamma': 2.5, 'dropout': 0.1}

Training convlstm | 6 epochs | device=cuda
AMP=True | accum_steps=1 | patience=6
focal_alpha=0.980 | focal_gamma=2.5
Expected gradient balance: fire=0.980 × N_fire ≈ no-fire=0.020 × N_no-fire


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   1/6 | Loss: 0.4305/0.4685 | Dice: 0.2249/0.0799 | Recall: 0.577 | LR: 5.00e-05 | 150.9s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   2/6 | Loss: 0.3899/0.4542 | Dice: 0.2999/0.0658 | Recall: 0.492 | LR: 1.00e-04 | 151.5s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   3/6 | Loss: 0.3678/0.4332 | Dice: 0.3470/0.1531 | Recall: 0.439 | LR: 1.00e-04 | 153.2s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   4/6 | Loss: 0.3611/0.4230 | Dice: 0.3654/0.2620 | Recall: 0.418 | LR: 8.55e-05 | 153.9s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   5/6 | Loss: 0.3593/0.4198 | Dice: 0.3730/0.1668 | Recall: 0.437 | LR: 5.05e-05 | 153.3s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   6/6 | Loss: 0.3568/0.4212 | Dice: 0.3818/0.2057 | Recall: 0.422 | LR: 1.55e-05 | 154.5s
Best val Dice: 0.2620
    Best val Dice: 0.2620 (epoch 4)

  Trial 3/3: {'learning_rate': 0.0005, 'weight_decay': 0.0001, 'focal_alpha': 0.995, 'focal_gamma': 3.0, 'dropout': 0.2}

Training convlstm | 6 epochs | device=cuda
AMP=True | accum_steps=1 | patience=6
focal_alpha=0.995 | focal_gamma=3.0
Expected gradient balance: fire=0.995 × N_fire ≈ no-fire=0.005 × N_no-fire


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   1/6 | Loss: 0.3879/0.4310 | Dice: 0.2713/0.2155 | Recall: 0.478 | LR: 2.50e-04 | 155.0s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   2/6 | Loss: 0.3647/0.4214 | Dice: 0.3733/0.2392 | Recall: 0.433 | LR: 5.00e-04 | 151.6s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   3/6 | Loss: 0.3578/0.4188 | Dice: 0.3798/0.2519 | Recall: 0.423 | LR: 5.00e-04 | 152.2s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   4/6 | Loss: 0.3555/0.4152 | Dice: 0.3904/0.2261 | Recall: 0.432 | LR: 4.27e-04 | 154.5s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   5/6 | Loss: 0.3537/0.4112 | Dice: 0.3886/0.2402 | Recall: 0.430 | LR: 2.51e-04 | 154.5s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   6/6 | Loss: 0.3494/0.4099 | Dice: 0.3972/0.2464 | Recall: 0.422 | LR: 7.41e-05 | 153.9s
Best val Dice: 0.2519
    Best val Dice: 0.2519 (epoch 3)

  ★ Best HP: {'learning_rate': 0.0003, 'weight_decay': 0.0001, 'focal_alpha': 0.99, 'focal_gamma': 2.0, 'dropout': 0.15} (Dice=0.2665)

  Final training: 30 epochs...

Training convlstm | 30 epochs | device=cuda
AMP=True | accum_steps=1 | patience=8
focal_alpha=0.990 | focal_gamma=2.0
Expected gradient balance: fire=0.990 × N_fire ≈ no-fire=0.010 × N_no-fire


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   1/30 | Loss: 0.4105/0.4548 | Dice: 0.1872/0.1370 | Recall: 0.523 | LR: 1.00e-04 | 2645.3s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   2/30 | Loss: 0.3719/0.4304 | Dice: 0.3327/0.1344 | Recall: 0.503 | LR: 2.00e-04 | 150.2s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   3/30 | Loss: 0.3620/0.4199 | Dice: 0.3780/0.1602 | Recall: 0.446 | LR: 3.00e-04 | 153.6s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   4/30 | Loss: 0.3591/0.4218 | Dice: 0.3833/0.1794 | Recall: 0.448 | LR: 3.00e-04 | 152.3s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   5/30 | Loss: 0.3556/0.4146 | Dice: 0.3902/0.2426 | Recall: 0.435 | LR: 2.99e-04 | 153.3s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   6/30 | Loss: 0.3541/0.4154 | Dice: 0.3964/0.2065 | Recall: 0.442 | LR: 2.96e-04 | 156.6s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   7/30 | Loss: 0.3520/0.4183 | Dice: 0.3976/0.2274 | Recall: 0.445 | LR: 2.91e-04 | 153.9s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   8/30 | Loss: 0.3496/0.4164 | Dice: 0.4055/0.2051 | Recall: 0.439 | LR: 2.84e-04 | 151.9s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch   9/30 | Loss: 0.3471/0.4134 | Dice: 0.4073/0.2611 | Recall: 0.434 | LR: 2.75e-04 | 153.7s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch  10/30 | Loss: 0.3465/0.4142 | Dice: 0.4126/0.2332 | Recall: 0.432 | LR: 2.65e-04 | 155.3s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch  11/30 | Loss: 0.3458/0.4154 | Dice: 0.4095/0.2314 | Recall: 0.436 | LR: 2.53e-04 | 154.9s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch  12/30 | Loss: 0.3451/0.4048 | Dice: 0.4150/0.2230 | Recall: 0.455 | LR: 2.40e-04 | 153.8s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch  13/30 | Loss: 0.3438/0.4094 | Dice: 0.4093/0.2242 | Recall: 0.427 | LR: 2.25e-04 | 153.2s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch  14/30 | Loss: 0.3431/0.4107 | Dice: 0.4183/0.2345 | Recall: 0.434 | LR: 2.10e-04 | 150.1s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch  15/30 | Loss: 0.3424/0.4077 | Dice: 0.4136/0.2523 | Recall: 0.427 | LR: 1.93e-04 | 151.4s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch  16/30 | Loss: 0.3426/0.4123 | Dice: 0.4149/0.2442 | Recall: 0.427 | LR: 1.76e-04 | 151.8s


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

Eval:   0%|          | 0/235 [00:00<?, ?it/s]

Epoch  17/30 | Loss: 0.3422/0.4076 | Dice: 0.4190/0.2277 | Recall: 0.433 | LR: 1.59e-04 | 152.5s
Early stopping at epoch 17 (best Dice=0.2611)
Best val Dice: 0.2611


c:\Users\Syrin\OneDrive\Bureau\FireForest\venv311\Lib\site-packages\torch\_utils.py:831: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
  return self.fget.__get__(instance, owner)()


  Optimal threshold: 0.85 (val Dice=0.2690)

══════════════════════════════════════════════════════════════════════
  MODEL: U-Net + Attention
══════════════════════════════════════════════════════════════════════
  WeightedSampler: 13273/14979 samples with fire | mean fire density 1.073%

  Trial 1/3: {'learning_rate': 0.0003, 'weight_decay': 0.0001, 'focal_alpha': 0.99, 'focal_gamma': 2.0, 'dropout': 0.15}

Training unet | 6 epochs | device=cuda
AMP=True | accum_steps=1 | patience=6
focal_alpha=0.990 | focal_gamma=2.0
Expected gradient balance: fire=0.990 × N_fire ≈ no-fire=0.010 × N_no-fire


Train:   0%|          | 0/1872 [00:00<?, ?it/s]

## 3.5 Results Summary

In [ ]:
# ══════════════════════════════════════════════════════════════
# Cell 6 — Results table (publication-ready)
# ══════════════════════════════════════════════════════════════
import pandas as pd

# Build results table
rows = []
for name in ['ca', 'convlstm', 'unet', 'pi_cca']:
    if name not in all_results:
        continue
    r = all_results[name]
    test_m = r['eval'].get('test', r['eval'].get('val', {}))
    val_m = r['eval'].get('val', {})
    rows.append({
        'Model': MODEL_CONFIG[name]['name'],
        'Type': MODEL_CONFIG[name].get('type', ''),
        'Threshold': f"{r['eval'].get('optimal_threshold', 0.5):.2f}",
        'Val Dice': f"{val_m.get('dice', 0):.4f}",
        'Val IoU': f"{val_m.get('iou', 0):.4f}",
        'Test Dice': f"{test_m.get('dice', 0):.4f}",
        'Test IoU': f"{test_m.get('iou', 0):.4f}",
        'Test F1': f"{test_m.get('f1', 0):.4f}",
        'Precision': f"{test_m.get('precision', 0):.4f}",
        'Recall': f"{test_m.get('recall', 0):.4f}",
    })

results_df = pd.DataFrame(rows)
print('\n' + '='*80)
print('  HOLDOUT TEST SET RESULTS — GTX 1050 Training')
print('='*80)
display(results_df)

# Save
results_df.to_csv(RESULTS_DIR / 'test_results_local.csv', index=False)
with open(RESULTS_DIR / 'all_results_local.json', 'w') as f:
    json.dump({k: v['eval'] for k, v in all_results.items()}, f, indent=2, default=str)

print(f'\nResults saved to {RESULTS_DIR}')

## 3.6 Training Curves

In [ ]:
# ══════════════════════════════════════════════════════════════
# Cell 7 — Training curves
# ══════════════════════════════════════════════════════════════

colors = {'convlstm': '#2196F3', 'unet': '#4CAF50', 'pi_cca': '#FF5722'}
labels = {k: MODEL_CONFIG[k]['name'] for k in colors}

fig, axes = plt.subplots(1, 3, figsize=(18, 5))

for name in candidate_models:
    if name not in all_results or 'history' not in all_results[name]:
        continue
    h = all_results[name]['history']
    c = colors[name]
    lbl = labels[name]
    epochs = range(1, len(h['train_loss']) + 1)

    # Loss
    axes[0].plot(epochs, h['train_loss'], color=c, linestyle='-', alpha=0.7, label=f'{lbl} (train)')
    axes[0].plot(epochs, h['val_loss'], color=c, linestyle='--', label=f'{lbl} (val)')

    # Dice
    val_dice = [m['dice'] for m in h['val_metrics']]
    train_dice = [m['dice'] for m in h['train_metrics']]
    axes[1].plot(epochs, train_dice, color=c, linestyle='-', alpha=0.7, label=f'{lbl} (train)')
    axes[1].plot(epochs, val_dice, color=c, linestyle='--', label=f'{lbl} (val)')

    # LR schedule
    axes[2].plot(epochs, h['lr'], color=c, label=lbl)

axes[0].set_xlabel('Epoch'); axes[0].set_ylabel('Loss (Focal + Dice)')
axes[0].set_title('Training & Validation Loss', fontweight='bold')
axes[0].legend(fontsize=7)

axes[1].set_xlabel('Epoch'); axes[1].set_ylabel('Dice Score')
axes[1].set_title('Dice Score', fontweight='bold')
axes[1].legend(fontsize=7)

axes[2].set_xlabel('Epoch'); axes[2].set_ylabel('Learning Rate')
axes[2].set_title('LR Schedule (Warmup + Cosine)', fontweight='bold')
axes[2].legend(fontsize=8)
axes[2].set_yscale('log')

plt.tight_layout()
plt.savefig(FIGURES_DIR / 'training_curves_local.png', dpi=200, bbox_inches='tight')
plt.show()

## 3.7 Qualitative Predictions

In [ ]:
# ══════════════════════════════════════════════════════════════
# Cell 8 — Visual prediction comparison
# ══════════════════════════════════════════════════════════════

# Load test batch
x_test, y_test = next(iter(loaders['test']))
x_test = x_test.to(device)

n_models = len(candidate_models)
n_samples = min(4, x_test.shape[0])

fig, axes = plt.subplots(n_samples, n_models + 2, figsize=(4*(n_models+2), 4*n_samples))
if n_samples == 1:
    axes = axes[np.newaxis, :]

for row in range(n_samples):
    # Input fire mask
    fire_in = x_test[row, CH['prev_fire_mask']].cpu().numpy()
    axes[row, 0].imshow(fire_in, cmap='hot', vmin=0, vmax=max(1, fire_in.max()))
    if row == 0: axes[row, 0].set_title('Input Fire', fontweight='bold')
    axes[row, 0].axis('off')

    # Ground truth
    gt = y_test[row].squeeze().numpy()
    axes[row, 1].imshow(gt, cmap='hot', vmin=0, vmax=1)
    if row == 0: axes[row, 1].set_title('Ground Truth', fontweight='bold')
    axes[row, 1].axis('off')

    # Predictions per model
    for col, mname in enumerate(candidate_models):
        ckpt = MODELS_DIR / mname / 'best_model.pt'
        if not ckpt.exists():
            continue
        model = MODEL_CLASSES[mname](config=MODEL_CONFIG[mname]).to(device)
        model.load_state_dict(torch.load(ckpt, map_location=device, weights_only=True))
        model.eval()

        opt_t = all_results.get(mname, {}).get('eval', {}).get('optimal_threshold', 0.5)
        with torch.no_grad():
            if torch.cuda.is_available():
                with autocast('cuda'):
                    pred = model(x_test[row:row+1])
            else:
                pred = model(x_test[row:row+1])
        pred_np = pred.float().squeeze().cpu().numpy()

        axes[row, col+2].imshow(pred_np > opt_t, cmap='hot', vmin=0, vmax=1)
        if row == 0:
            axes[row, col+2].set_title(f'{MODEL_CONFIG[mname]["name"]}\n(t={opt_t:.2f})', fontweight='bold', fontsize=9)
        axes[row, col+2].axis('off')
        del model

plt.suptitle('Test Set Predictions — GTX 1050 Training', fontsize=14, fontweight='bold', y=1.01)
plt.tight_layout()
plt.savefig(FIGURES_DIR / 'predictions_local.png', dpi=200, bbox_inches='tight')
plt.show()

if torch.cuda.is_available():
    torch.cuda.empty_cache()

## 3.8 Threshold Sensitivity Analysis

In [ ]:
# ══════════════════════════════════════════════════════════════
# Cell 9 — Threshold sensitivity (important for publication)
# ══════════════════════════════════════════════════════════════

thresholds = np.arange(0.05, 0.95, 0.05)

fig, axes = plt.subplots(1, 2, figsize=(14, 5))

for mname in candidate_models:
    ckpt = MODELS_DIR / mname / 'best_model.pt'
    if not ckpt.exists():
        continue

    model = MODEL_CLASSES[mname](config=MODEL_CONFIG[mname]).to(device)
    model.load_state_dict(torch.load(ckpt, map_location=device, weights_only=True))
    model.eval()

    # Collect val predictions
    val_preds, val_targets = [], []
    with torch.no_grad():
        for x, y in loaders['val']:
            x = x.to(device)
            if torch.cuda.is_available():
                with autocast('cuda'):
                    pred = model(x)
            else:
                pred = model(x)
            val_preds.append(pred.float().cpu())
            val_targets.append(y)

    val_preds = torch.cat(val_preds)
    val_targets = torch.cat(val_targets)

    dices, ious = [], []
    for t in thresholds:
        m = compute_metrics_at_threshold(val_preds, val_targets, float(t))
        dices.append(m['dice'])
        ious.append(m['iou'])

    c = colors[mname]
    axes[0].plot(thresholds, dices, color=c, marker='.', label=labels[mname])
    axes[1].plot(thresholds, ious, color=c, marker='.', label=labels[mname])

    del model
    if torch.cuda.is_available():
        torch.cuda.empty_cache()

axes[0].set_xlabel('Threshold'); axes[0].set_ylabel('Dice')
axes[0].set_title('Dice vs Threshold (Validation)', fontweight='bold')
axes[0].legend()
axes[0].grid(True, alpha=0.3)

axes[1].set_xlabel('Threshold'); axes[1].set_ylabel('IoU')
axes[1].set_title('IoU vs Threshold (Validation)', fontweight='bold')
axes[1].legend()
axes[1].grid(True, alpha=0.3)

plt.tight_layout()
plt.savefig(FIGURES_DIR / 'threshold_sensitivity_local.png', dpi=200, bbox_inches='tight')
plt.show()

## Summary

### Improvements over previous FastTrack notebook

1. **Full dataset used** — no subsampling (14,979 train / 1,877 val / 1,689 test)
2. **Mixed precision (AMP)** — ~2× speedup, ~50% VRAM reduction
3. **All 4 models trained** including PI-CCA (project's novel contribution)
4. **Proper warmup + cosine decay** LR schedule
5. **3 structured HP trials** per model
6. **40 epoch budget** with early stopping (patience=10)
7. **Optimal threshold** per model via validation sweep
8. **Gradient accumulation** for PI-CCA (effective bs=8 with actual bs=4)
9. **Publication-ready visualisations** with DPI=200

### Limitations (local GTX 1050)

- Limited VRAM prevents larger batch sizes
- Fewer tuning trials than desirable for a full ablation
- No test-time augmentation (TTA)

For Camber cloud execution with more resources, see **03_Training_Camber.ipynb**.